### ISSUE TITLE
Error reporting: Compiling and using both Matlab and Python

### ISSUE THREAD

**Issue by inwoo-park** [ISSMteam/ISSM#27]
Hi, Developers.

I apologize for raising another issue, which stems from [GitHub issue #17](https://github.com/ISSMteam/ISSM/issues/17#issue-2516594400)

# ISSUE

I usually use the ISSM with both python and matlab. Therefore, I compile the ISSM with python and matlab as follows
(I think that ISSM does not recommand to compile both Python and MATLAB...)

```bash
./configure ... --with-matlab=$MATLAB_DIR --with-python=$PYTHON_DIR
```

After issuing the the problem with using Python [GitHub issue #17](https://github.com/ISSMteam/ISSM/issues/17#issue-2516594400), the solution is accepted and fix the problem.

However, after fixing the `src/wrappers/python/Makefile.am`. I encounter the another issue as follows:

(**NOTE**: I translate error messages from Korean to English, so the English version may not be an exact match.)

```bash
>> test101
Invalid mex file in '/home/inwoo/ISSM/mpich-petsc3.19-python3.11/lib/IssmConfig_matlab.mexa64':
/home/inwoo/ISSM/mpich-petsc3.19-python3.11/lib/libISSMApi.so: undefined symbol: PyCapsule_Type.

ERROR: IssmConfig (line 14)
value = IssmConfig_matlab(string);

ERROR: toolkits/setdefaultparameters (line 57)
                        if IssmConfig('_HAVE_PETSC_'),

ERROR: toolkits (line 27)
                                        self=setdefaultparameters(self);

ERROR: model/setdefaultparameters (line 283)
                        md.toolkits         = toolkits();

ERROR: model (line 206)
                                        md=setdefaultparameters(md,'earth');

ERROR: test101 (line 2)
md=triangle(model(),'../Exp/Square.exp',50000.);
```

# What is the problem?

I think that the name of `libISSMApi.so` for MATLAB and Python may trigger the above error. It seems that when the `IssmConfig_matlab.mexa64` file attempts to load `libISSMApi.so`, there is confusion between the Python and MATLAB versions.

`src/wrappers/python/Maefile.am` Line 27-36

```bash
#api io{{{
lib_LTLIBRARIES += libISSMApi.la

if !MSYS2
api_sources= ./io/ApiPrintf.cpp

libISSMApi_la_SOURCES = $(api_sources)
libISSMApi_la_CXXFLAGS = ${ALL_CXXFLAGS}
endif
```

`src/wrappers/matlab/Makefile.am` Line 29-38

```bash
#api io{{{
lib_LTLIBRARIES += libISSMApi.la

if !MSYS2
api_sources= ./io/ApiPrintf.cpp

libISSMApi_la_SOURCES = $(api_sources)
libISSMApi_la_CXXFLAGS = ${ALL_CXXFLAGS}
endif
#}}}
```

Before resolving [GitHub issue #17](https://github.com/ISSMteam/ISSM/issues/17#issue-2516594400), `libISSMApi.so` behaved identically in both MATLAB and Python. However, after resolving the issue, `libISSMApi.so` in MATLAB and Python no longer function the same way, leading to the above error.

# Solution

Using different name in `libISSMApi.so` for Python and MATLAB may resolve above problem.

How about changing `libISSMApi` to `libISSMApi_Python` in `src/wrappers/python/Makefile.am`? I performed a quick check on the following changes and confirmed that they are working correctly. 

Before

`lib_LTLIBRARIES += libISSMApi.la`

Changed

`lib_LTLIBRARIES += libISSMApi_Python.la`

```bash
#api io{{{
lib_LTLIBRARIES += libISSMApi_Python.la

if !MSYS2
api_sources= ./io/ApiPrintf.cpp

libISSMApi_la_SOURCES = $(api_sources)
libISSMApi_la_CXXFLAGS = ${ALL_CXXFLAGS}
endif
```

Best regards,

Inwoo

**Comment by MathieuMorlighem** [comment]
weird, I have compiled ISSM with both MATLAB and python in the past and it has always worked just fine. Let me try again

**Comment by MathieuMorlighem** [comment]
Yes, works on my end (although I had to fix a bug in `issm_options.m4` because ISSM was not picking up the right version of python).

**Comment by inwoo-park** [comment]
Hi, Mathieu

After pulling your fix and compiling the latest ISSM, I got the same issue in local linux machine. 

I have two machine, one for local (got problem) and the other for cluster (server).
The server works well, while the local machine has a same problem...

Hmm.... 

I suspect that both python and matlab wrappers contain `libISSMApi.so` for `IssmConfig_matlab.mexa64` and `IssmConfig_python.so`, respectively. Therefore, Would we clarify where the `lib/libISSMApi.so` originates from? 

* On linux machine (local machine)

```bash
>> test101
Invalid MEX mex '/home/inwoo/ISSM/mpich-petsc3.19-python3.11/lib/IssmConfig_matlab.mexa64'임:
/home/inwoo/ISSM/mpich-petsc3.19-python3.11/lib/libISSMApi.so: undefined symbol: PyCapsule_Type.

ERROR: IssmConfig (line 14)
value = IssmConfig_matlab(string);

ERROR: toolkits/setdefaultparameters (line 57)
                        if IssmConfig('_HAVE_PETSC_'),

ERROR: toolkits (line 27)
                                        self=setdefaultparameters(self);

ERROR: model/setdefaultparameters (line 283)
                        md.toolkits         = toolkits();

ERROR: model (line 206)
                                        md=setdefaultparameters(md,'earth');

ERROR: test101 (line 2)
md=triangle(model(),'../Exp/Square.exp',50000.);
```


* On Server machine (cluster)

```bash
>> test101
      boundary conditions for stressbalance model: spc set as zero
      no balancethickness.thickening_rate specified: values set as zero
uploading input file and queuing script
launching solution sequence on remote cluster

Ice-sheet and Sea-level System Model (ISSM) version  4.24
(website: http://issm.jpl.nasa.gov forum: https://issm.ess.uci.edu/forum/)

call computational core:
write lock file:

   FemModel initialization elapsed time:   0.0165768
   Total Core solution elapsed time:       0.0634174
   Linear solver elapsed time:             0.0542367 (86%)

   Total elapsed time: 0 hrs 0 min 0 sec
```

**Comment by TidbitSoftware** [comment]
Hi guys,
Traditionally I have had no issue with compiling both APIs and running either as needed. This applied to both Linux and macOS. Inwoo, it looks like it's now working on your server, but "On linux machine (local machine)" it now is not? Which flavor of Linux is this machine running? Your idea to fix the issue by splitting the library into two will work, but I want to try to reproduce the issue on my end.

**Comment by inwoo-park** [comment]
Hi,

My local machine and server machine linux version are 

Local                - Ubuntu 16.04.7 LTS (Intel(R) Xeon(R) CPU E3-1231 v3)
Server(cluster)  - 

The configuration for local machine is 
```
./configure --prefix=/home/inwoo/ISSM/mpich-petsc3.19-python3.11 \
--with-wrappers=yes \
--with-bamg=yes \--with-kriging=yes --with-matlab-dir=/home/inwoo/externalpackages/matlab/install --with-mpi-libdir=-L/home/inwoo/externalpackages/mpich/dev/lib -lmpi --with-mpi-include=/home/inwoo/externalpackages/mpich/dev/include --with-m1qn3-dir=/home/inwoo/ISSM/externalpackages/m1qn3/install --with-gsl-dir=/home/inwoo/externalpackages/gsl/2.6 --with-gmsh-dir=/home/inwoo/externalpackages/gmsh/4.8 --with-triangle-dir=/home/inwoo/ISSM/externalpackages/triangle/install --with-math77-dir=/home/inwoo/ISSM/externalpackages/math77/install/ --with-gmsh-dir=/home/inwoo/externalpackages/gmsh/4.8 --with-metis-dir=/home/inwoo/externalpackages/petsc/3.19 --with-blas-lapack-dir=/home/inwoo/externalpackages/petsc/3.19 --with-mumps-dir=/home/inwoo/externalpackages/petsc/3.19 --with-scalapack-dir=/home/inwoo/externalpackages/petsc/3.19 --with-petsc-dir=/home/inwoo/externalpackages/petsc/3.19 --enable-development --enable-standalone-executables --with-ocean=yes --with-fortran-lib=-L/usr/lib/x86_64-linux-gnu/ -lgfortran --with-semic-dir=/home/inwoo/ISSM/externalpackages/semic/install --with-python-dir=/home/inwoo/externalpackages/python/3.11/bin/../ --with-python-version=3.11 --with-python-numpy-dir=/home/inwoo/externalpackages/python/3.11/lib/python3.11/site-packages/numpy --with-numthreads=8 --enable-shared=yes --enable-static=no --with-64bit-indices=0 CXXFLAGS= -I/home/inwoo/externalpackages/mpich/dev/include -std=c++11 CFLAGS= -I/home/inwoo/externalpackages/mpich/dev/include -std=c++11 LDFLAGS=  -L/home/inwoo/externalpackages/python/3.11/lib  -lpthread -ldl  -lutil -lm  -lstdc++ LIBS= CC=gcc CXX=g++ FC=gfortran
```

Best regards,
Inwoo

**Comment by inwoo-park** [comment]
Hi,

Here is some information about my local machine and server machine configurations:

Local Machine:
* OS: Ubuntu 16.04.7 LTS 
* CPU: Intel(R) Xeon(R) CPU E3-1231 v3
* Compiler: GCC 9.4.0

Server(Cluster) Machine:
* OS: CentOS Linux 7 
* CPU: Intel(R) Xeon(R) Gold 6150
* Compiler: Interl compiler (icc (ICC) 18.0.3 20180410)

I usually build Python from source instead of using the pre-compiled Python available via `sudo apt-get install python`. As a result, Python is not located in the `/usr` directory.

The configuration for local machine is 
```
./configure --prefix=/home/inwoo/ISSM/mpich-petsc3.19-python3.11 
--with-wrappers=yes 
--with-bamg=yes 
--with-kriging=yes 
--with-matlab-dir=/home/inwoo/externalpackages/matlab/install 
--with-mpi-libdir=-L/home/inwoo/externalpackages/mpich/dev/lib -lmpi --with-mpi-include=/home/inwoo/externalpackages/mpich/dev/include 
--with-m1qn3-dir=/home/inwoo/ISSM/externalpackages/m1qn3/install 
--with-gsl-dir=/home/inwoo/externalpackages/gsl/2.6 
--with-gmsh-dir=/home/inwoo/externalpackages/gmsh/4.8 
--with-triangle-dir=/home/inwoo/ISSM/externalpackages/triangle/install 
--with-math77-dir=/home/inwoo/ISSM/externalpackages/math77/install/ 
--with-metis-dir=/home/inwoo/externalpackages/petsc/3.19 
--with-blas-lapack-dir=/home/inwoo/externalpackages/petsc/3.19 
--with-mumps-dir=/home/inwoo/externalpackages/petsc/3.19 
--with-scalapack-dir=/home/inwoo/externalpackages/petsc/3.19 
--with-petsc-dir=/home/inwoo/externalpackages/petsc/3.19 
--enable-development 
--enable-standalone-executables 
--with-ocean=yes 
--with-fortran-lib=-L/usr/lib/x86_64-linux-gnu/ -lgfortran 
--with-semic-dir=/home/inwoo/ISSM/externalpackages/semic/install 
--with-python-dir=/home/inwoo/externalpackages/python/3.11/bin/../ 
--with-python-version=3.11 
--with-python-numpy-dir=/home/inwoo/externalpackages/python/3.11/lib/python3.11/site-packages/numpy 
--with-numthreads=8 
--enable-shared=yes --enable-static=yes
--with-64bit-indices=0 CXXFLAGS= -I/home/inwoo/externalpackages/mpich/dev/include -std=c++11 
CFLAGS= -I/home/inwoo/externalpackages/mpich/dev/include -std=c++11 
LDFLAGS=  -L/home/inwoo/externalpackages/python/3.11/lib  -lpthread -ldl  -lutil -lm  -lstdc++ 
LIBS= CC=gcc CXX=g++ FC=gfortran
```

The "configuration" in server machine is 
```
./configure --prefix=/home/inwoo/issm/ISSM/mpich-petsc3.18-python3.11
 --with-wrappers=yes
 --with-bamg=yes 
 --with-kriging=yes 
 --with-matlab-dir=/home/DATA/externalpackages/matlab/install
 --with-mpi-libdir=-L/usr/local/mpi/intel18/mpich-3.2//lib -lmpi
 --with-mpi-include=/usr/local/mpi/intel18/mpich-3.2//include
 --with-m1qn3-dir=/home/inwoo/issm/ISSM/externalpackages/m1qn3/install
 --with-gsl-dir=/home/DATA/externalpackages/gsl/2.6
 --with-chaco-dir=/home/DATA/externalpackages/chaco
 --with-triangle-dir=/home/inwoo/issm/ISSM/externalpackages/triangle/install
 --with-math77-dir=/home/DATA/externalpackages/math77
 --with-gmsh-dir=/home/DATA/externalpackages/gmsh/4.4
 --with-metis-dir=/home/DATA/externalpackages/petsc/intel-mpich/3.18
 --with-blas-lapack-dir=/home/DATA/externalpackages/petsc/intel-mpich/3.18
 --with-mumps-dir=/home/DATA/externalpackages/petsc/intel-mpich/3.18
 --with-scalapack-dir=/home/DATA/externalpackages/petsc/intel-mpich/3.18
 --with-petsc-dir=/home/DATA/externalpackages/petsc/intel-mpich/3.18 
 --enable-development
 --enable-standalone-executables 
 --with-ocean=yes 
 --with-fortran-lib=-L/usr/lib64 -lgfortran
 --with-semic-dir=/home/inwoo/issm/ISSM/externalpackages/semic/install --with-python-dir=/home/DATA/externalpackages/python/3.11/bin/../ 
 --with-python-version=3.11 --with-python-numpy-dir=/home/DATA/externalpackages/python/3.11/lib/python3.11/site-packages/numpy --with-numthreads=36 
 --enable-shared=yes --enable-static=yes --with-64bit-indices=0 
 CXXFLAGS= -I/usr/local/mpi/intel18/mpich-3.2//include -std=c++11 
 CFLAGS= -I/usr/local/mpi/intel18/mpich-3.2//include -std=c++11 
 LDFLAGS= -L/usr/local/intel/lib/intel64 
 LIBS= -lifport -lifcoremt 
 CC=mpicc CXX=mpic++ FC=mpifort
```

Best regards,
Inwoo

**Comment by TidbitSoftware** [comment]
Inwoo,
As an aside, you should remove the following from your configurations,
```
--enable-shared=yes
--enable-static=yes
--enable-standalone-executables
```
and add,
```
--disable-static
```

**Comment by inwoo-park** [comment]
Hi, @TidbitSoftware 

Thank you for your recommendtation!

On the server, sometimes, I check the test files while preparing model setups. Is it okay to keep `--with-wrapers=yes` enabled on the server?

Thanks.

**Comment by TidbitSoftware** [comment]
Hi @inwoo-park,
In this case, keep the `--with-wrappers` and MATLAB- and Python-related options. I was suggesting that just to simplify things if you weren't using them on the remote server.

**Comment by inwoo-park** [comment]
Hi @TidbitSoftware,

Thanks!

And is this issue scheduled to be closed?

**Comment by TidbitSoftware** [comment]
Hi Inwoo,
I will not close it yet unless you can confirm that either (1) the changing of the options on your local machine (e.g. removing the `--enable-static` option) resolves the issue, or, (2) I can test out the PR you've made to see exactly why it's working in on other platforms but not in your use case. I was not able to reproduce the issue today on my local Mac machine but do want to see still if I can under another Ubuntu-based machine.

**Comment by inwoo-park** [comment]
Hi, @TidbitSoftware 

Without changing `Makefile.am`.

Current git commit.
```bash
commit a178c1069fbba3a0c7adcb3f42932407d45d6ab6 (HEAD -> main, origin/main, origin/HEAD, fix-makefile)
Author: Tidbit Software <support@tidbitsoftware.com>
Date:   Mon Nov 11 20:00:28 2024 -0800

    CHG: Moving certain external packages to new repo
```

Here is the configuration log file (`config.log`) on Local machine (not server).  I set the configuration with `--disable-static`.

1. I will check the other local machine (Ubuntu 20.04 machine).

2. Now, this old machine (Ubuntu 16.04) is not my main machine. Therefore, this issue may not be the problem. I made it difficult.


```bash
./configure --prefix=/home/inwoo/ISSM/mpich-petsc3.19-python3.11 
 --with-wrappers=yes
 --with-bamg=yes
 --with-kriging=yes --with-matlab-dir=/home/inwoo/externalpackages/matlab/install
 --with-mpi-libdir=-L/home/inwoo/externalpackages/mpich/dev/lib -lmpi
 --with-mpi-include=/home/inwoo/externalpackages/mpich/dev/include
 --with-m1qn3-dir=/home/inwoo/ISSM/externalpackages/m1qn3/install
 --with-gsl-dir=/home/inwoo/externalpackages/gsl/2.6
 --with-triangle-dir=/home/inwoo/ISSM/externalpackages/triangle/install
 --with-math77-dir=/home/inwoo/ISSM/externalpackages/math77/install/
 --with-gmsh-dir=/home/inwoo/externalpackages/gmsh/4.8
 --with-metis-dir=/home/inwoo/externalpackages/petsc/3.19
 --with-blas-lapack-dir=/home/inwoo/externalpackages/petsc/3.19
 --with-mumps-dir=/home/inwoo/externalpackages/petsc/3.19
 --with-scalapack-dir=/home/inwoo/externalpackages/petsc/3.19
 --with-petsc-dir=/home/inwoo/externalpackages/petsc/3.19
 --enable-development --with-ocean=yes 
 --with-fortran-lib=-L/usr/lib/x86_64-linux-gnu/ -lgfortran
 --with-semic-dir=/home/inwoo/ISSM/externalpackages/semic/install
 --with-python-dir=/home/inwoo/externalpackages/python/3.11/bin/../ 
--with-python-version=3.11
--with-python-numpy-dir=/home/inwoo/externalpackages/python/3.11/lib/python3.11/site-packages/numpy
--with-numthreads=8
--disable-static
--with-64bit-indices=0
CXXFLAGS= -I/home/inwoo/externalpackages/mpich/dev/include -std=c++11 CFLAGS= -I/home/inwoo/externalpackages/mpich/dev/include -std=c++11 
LDFLAGS=  -L/home/inwoo/externalpackages/python/3.11/lib  -lpthread -ldl  -lutil -lm  -lstdc++ LIBS= CC=gcc CXX=g++ FC=gfortran
```

And log message... So sad... 

```
>> test101
invalid MEX file '/home/inwoo/ISSM/mpich-petsc3.19-python3.11/lib/IssmConfig_matlab.mexa64'임:
/home/inwoo/ISSM/mpich-petsc3.19-python3.11/lib/libISSMApi.so: undefined symbol: PyCapsule_Type.

ERROR: IssmConfig (line 14)
value = IssmConfig_matlab(string);

ERROR: toolkits/setdefaultparameters (line 57)
                        if IssmConfig('_HAVE_PETSC_'),

ERROR: toolkits (line 27)
                                        self=setdefaultparameters(self);

ERROR: model/setdefaultparameters (line 283)
                        md.toolkits         = toolkits();

ERROR: model (line 206)
                                        md=setdefaultparameters(md,'earth');

ERROR: test101 (line 2)
md=triangle(model(),'../Exp/Square.exp',50000.);
```

Thanks..

**Comment by TidbitSoftware** [comment]
Hi Inwoo,
Thanks for trying this out anyhow! I will work on testing and integrating your changes over the next day or so.

**Comment by TidbitSoftware** [comment]
Hi Inwoo,
I've tested and merged your changes with some minor adjustments, so I'll be closing this PR but your changes are committed. Thanks!

**Comment by inwoo-park** [comment]
Hi @TidbitSoftware,

That is a great news!
Thanks!

Have a nice day!
Inwoo.
